pkg/storegateway/indexheader/index/postings.go

```diff
@@ -409,22 +441,87 @@ func (t *PostingOffsetTableV2) PostingsOffset(name string, value string) (r inde
 	return index.Range{}, false, nil
 }
 
-func (t *PostingOffsetTableV2) LabelValues(name string, prefix string, filter func(string) bool) (v []string, err error) {
+var labelValuesAccumulators = sync.Pool{New: func() any {
+	return &postingsTableV2Accumulator[string]{
+		transform: func(offset PostingListOffset) string {
+			return offset.Value
+		},
+	}
+}}
+
+func (t *PostingOffsetTableV2) LabelValues(name string, prefix string, filter func(string) bool) ([]string, error) {
+	acc := labelValuesAccumulators.Get().(*postingsTableV2Accumulator[string])
+	defer func() {
+		acc.acc = nil
+		labelValuesAccumulators.Put(acc)
+	}()
+
+	if err := t.postingOffsets(name, prefix, filter, acc); err != nil {
+		return nil, errors.Wrap(err, "get label values")
+	}
+	return acc.acc, nil
+}
+
+var labelValuesOffsetsAccumulators = sync.Pool{New: func() any {
+	return &postingsTableV2Accumulator[PostingListOffset]{
+		transform: func(offset PostingListOffset) PostingListOffset {
+			return offset
+		},
+	}
+}}
+
+func (t *PostingOffsetTableV2) LabelValuesOffsets(name, prefix string, filter func(string) bool) ([]PostingListOffset, error) {
+	acc := labelValuesOffsetsAccumulators.Get().(*postingsTableV2Accumulator[PostingListOffset])
+	defer func() {
+		acc.acc = nil
+		labelValuesOffsetsAccumulators.Put(acc)
+	}()
+
+	if err := t.postingOffsets(name, prefix, filter, acc); err != nil {
+		return nil, errors.Wrap(err, "get label values offsets")
+	}
+	return acc.acc, nil
+}
+
+type postingsTableV2Accumulator[T any] struct {
+	transform func(PostingListOffset) T
+	acc       []T
+}
+
+func (v *postingsTableV2Accumulator[T]) estimatedValues(n int) {
+	v.acc = make([]T, 0, n)
+}
+
+func (v *postingsTableV2Accumulator[T]) visit(offset PostingListOffset) {
+	v.acc = append(v.acc, v.transform(offset))
+}
+
+type postingsTableV2Visitor interface {
+	estimatedValues(n int)
+	visit(offset PostingListOffset)
+}
+
+func (t *PostingOffsetTableV2) postingOffsets(name string, prefix string, filter func(string) bool, visitor postingsTableV2Visitor) (err error) {
```

Comment:
Calling an interface here for every single label value might be an unnecessary overhead.  I'd consider passing the two functions, or just making the entire method a generic function.

Reply:
i was surprised, but the calls even for the generic interface get inlined by the go compiler. I'll try to get you the compiler logs for the module later.

Reply:
I'd get rid of the interface (and of the visitor, and pools) by doing this: https://github.com/grafana/mimir/commit/39bfb954291b83335071316481da287eebb7e69b